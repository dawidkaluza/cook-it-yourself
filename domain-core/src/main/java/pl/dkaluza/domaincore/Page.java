package pl.dkaluza.domaincore;

import java.util.List;

import static pl.dkaluza.domaincore.Validator.validator;

public class Page<T> {
    private final List<T> items;
    private final int pageNumber;
    private final int totalPages;

    private Page(List<T> items, int pageNumber, int totalPages) {
        this.items = items;
        this.pageNumber = pageNumber;
        this.totalPages = totalPages;
    }

    public static <T> Factory<Page<T>> of(List<T> items) {
        return new PageFactory<>(items, 1, 1);
    }

    public static <T> Factory<Page<T>> of(List<T> items, int pageNumber, int totalPages) {
        return new PageFactory<>(items, pageNumber, totalPages);
    }

    public List<T> getItems() {
        return items;
    }

    public int getPageNumber() {
        return pageNumber;
    }

    public int getTotalPages() {
        return totalPages;
    }

    private static class PageFactory<T> extends DefaultFactory<Page<T>> {
        private PageFactory(List<T> items, int pageNumber, int totalPages) {
            super(
                ValidationExecutor.builder()
                    .withValidator(validator(items != null, "items", "Items must not be null"))
                    .withValidator(validator(pageNumber > 0, "pageNumber", "Page number must be a positive number"))
                    .withValidator(validator(totalPages > 0, "totalPages", "Total pages must be a positive number"))
                    .build(),
                () -> new Page<>(items, pageNumber, totalPages)
            );
        }
    }
}
