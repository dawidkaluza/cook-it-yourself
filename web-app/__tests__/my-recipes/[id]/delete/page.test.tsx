import {useFormState} from "react-dom";
import {Mock} from "vitest";
import {render, screen} from "@testing-library/react";
import Page from "@/app/my-recipes/[id]/delete/page";

vi.mock("react-dom", () => {
  return {
    useFormState: vi.fn(),
  };
});

describe("page component", () => {
  test.each([
    [ "" ],
    [ "The recipe does not exist." ],
  ])("render", (errorMessage) => {
    // Given
    const recipeId = 1;
    (useFormState as Mock).mockReturnValue([
      errorMessage,
      () => {}
    ]);

    // When
    const formComponent = render(<Page params={{ id: 1 }} />);

    // Then
    const questionText = screen.getByText(/are you sure/i);
    expect(questionText).not.toBeNull();

    const deleteButton = screen.getByRole("button", { name: /delete/i });
    expect(deleteButton).not.toBeNull();

    const cancelButton  = screen.getByRole("link", { name: /cancel/i });
    expect(cancelButton).not.toBeNull();

    if (errorMessage) {
      const errorText = screen.getByText(errorMessage);
      expect(errorText).not.toBeNull();
    }

    expect(formComponent.container).toMatchSnapshot();
  });
});