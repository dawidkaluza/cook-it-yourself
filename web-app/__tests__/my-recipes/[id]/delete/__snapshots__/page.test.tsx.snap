// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page component > render 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-12 text-center"
      >
        <h1>
          Delete the recipe
        </h1>
        <form
          novalidate=""
        >
          <input
            name="id"
            type="hidden"
            value="1"
          />
          <p>
            Are you sure that you want to delete the recipe?
          </p>
          <div
            class="my-4 d-flex justify-content-center"
          >
            <div
              class="btn-group"
            >
              <button
                class="btn btn-danger"
                type="submit"
              >
                <svg
                  class="bi bi-x"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
                  />
                </svg>
                 Delete 
              </button>
              <a
                class="btn btn-outline-secondary"
                href="/my-recipes/1"
              >
                Cancel
              </a>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`page component > render 2`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-12 text-center"
      >
        <h1>
          Delete the recipe
        </h1>
        <form
          novalidate=""
        >
          <input
            name="id"
            type="hidden"
            value="1"
          />
          <p>
            Are you sure that you want to delete the recipe?
          </p>
          <div
            class="my-4 d-flex justify-content-center"
          >
            <div
              class="btn-group"
            >
              <button
                class="btn btn-danger"
                type="submit"
              >
                <svg
                  class="bi bi-x"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
                  />
                </svg>
                 Delete 
              </button>
              <a
                class="btn btn-outline-secondary"
                href="/my-recipes/1"
              >
                Cancel
              </a>
            </div>
          </div>
          <p
            class="text-danger"
          >
            The recipe does not exist.
          </p>
        </form>
      </div>
    </div>
  </div>
</div>
`;
